### PR TITLE
fix Issue 23598 - Circular reference bug with static if and eponymous…

### PR DESCRIPTION
**Pull request by WalterBright**
… templates

### CODE REVIEW

**Comment by WalterBright** [comment]
It be really nice if the tester would run the compiler test suite before trying to build everything else first.

**Comment by RazvanN7** [comment]
The test suite is not running any tests. Druntime cannot be build with this patch.

**Comment by ibuclaw** [comment]
> It be really nice if the tester would run the compiler test suite before trying to build everything else first.

Can't run the compiler test suite without druntime.

**Comment by RazvanN7** [comment]
"

src/core/lifetime.d(130): Error: static assert:  "Don't know how to initialize an object of type AssertError with arguments (string, ulong)"
--
  | src/core/exception.d(873):        instantiated from here: `emplace!(AssertError, string, ulong)`
  | src/core/exception.d(556):        instantiated from here: `staticError!(AssertError, string, ulong)`

"

**Comment by WalterBright** [comment]
> Can't run the compiler test suite without druntime.

Not always, but nearly always one can, by using the previous druntime.

File: compiler/src/dmd/dtemplate.d [hunk 1]
```diff
@@ -7507,6 +7508,35 @@ extern (C++) class TemplateInstance : ScopeDsymbol
...
+             * Then try to set aliasdecl.
+             * Later do the if/else bodies.
+             * https://issues.dlang.org/show_bug.cgi?id=23598
+             */
+            void staticIfDg(Dsymbol s)
+            {
+                if (aliasdecl)
+                    return;
+                //printf("\t static if on '%s' %p kind %s in '%s'\n",  s.toChars(), s, s.kind(), this.toChars());
+                if (strcmp(s.kind(), "static if"))
```
[the remark this answers is not in the document]
Comment (BorisCarvajal) [reply compiler/src/dmd/dtemplate.d:7525]: Come on!
```d
diff --git a/compiler/src/dmd/attrib.d b/compiler/src/dmd/attrib.d
index 712099ed33..f5dd92fecc 100644
--- a/compiler/src/dmd/attrib.d
+++ b/compiler/src/dmd/attrib.d
@@ -1080,6 +1080,11 @@ extern (C++) final class StaticIfDeclaration : ConditionalDeclaration
         return "static if";
     }
 
+    override inout(StaticIfDeclaration) isStaticIfDeclaration() inout
+    {
+        return this;
+    }
+
     override void accept(Visitor v)
     {
         v.visit(this);
diff --git a/compiler/src/dmd/dsymbol.d b/compiler/src/dmd/dsymbol.d
index a17c830b6e..a9ca5b71d8 100644
--- a/compiler/src/dmd/dsymbol.d
+++ b/compiler/src/dmd/dsymbol.d
@@ -1413,6 +1413,7 @@ extern (C++) class Dsymbol : ASTNode
     inout(OverloadSet)                 isOverloadSet()                 inout { return null; }
     inout(CompileDeclaration)          isCompileDeclaration()          inout { return null; }
     inout(StaticAssert)                isStaticAssert()                inout { return null; }
+    inout(StaticIfDeclaration)         isStaticIfDeclaration()         inout { return null; }
 }
 
 /***********************************************************
```

**Review by ibuclaw — DISMISSED** [review]
Agreeing with @BorisCarvajal

File: compiler/src/dmd/dtemplate.d (hunk 1 above)
[the remark this answers is not in the document]
Comment (WalterBright) [reply compiler/src/dmd/dtemplate.d:7525]: There's more to it than that.

[the remark this answers is not in the document]
Comment (WalterBright) [reply compiler/src/dmd/dtemplate.d]: now that it's working, I made this change

**Comment by WalterBright** [comment]
The buildkite error appears to be here:
https://github.com/jmdavis/dxml/blob/master/source/dxml/parser.d#L3631

The lines don't quite line up with line 3628.

**Comment by maxsamukha** [comment]
This seems to have fixed the original issue, but it didn't take me long to hit another one, which will take a while to reduce.

**Comment by WalterBright** [comment]
Please file separate issues with separate entries in Bugzilla.

**Comment by maxsamukha** [comment]
> Please file separate issues with separate entries in Bugzilla.

If it proves to be a different issue and not another manifestation of this same issue.

**Comment by WalterBright** [comment]
If the original is fixed, then another manifestation is a different issue. Please do not append more issues to the same issue. It makes things very hard to manage.

**Comment by WalterBright** [comment]
Related issues can be associated by adding to the [See Also] section.

**Comment by maxsamukha** [comment]
> If the original is fixed, then another manifestation is a different issue. Please do not append more issues to the same issue. It makes things very hard to manage.

The fact that the original test case compiles doesn't necessarily mean the issue has been fixed. That's why I wrote "it seems to have been fixed". If it proves to be a different issue, I will file a different issue. If it's an incomplete or incorrect fix, I fill reopen this bug report.

**Comment by ibuclaw** [comment]
> The buildkite error appears to be here:
> https://github.com/jmdavis/dxml/blob/master/source/dxml/parser.d#L3631
> 
> The lines don't quite line up with line 3628.

You're looking at master, not the 0.4.3 release tag.

https://github.com/jmdavis/dxml/blob/d8eb8cff883d78a6258fc1e829484262c6f009fb/source/dxml/parser.d#L3628

**Comment by WalterBright** [comment]
> If it's an incomplete or incorrect fix, I will reopen this bug report.

I have a lot of experience with this. Please open a new report. The problem with reopening with something else is we lose the correspondence between issues and PRs. Also, what may appear to be the same problem may not actually be the same one.

**Comment by WalterBright** [comment]
Anyone care to have a look at jmdavis' project to see what the actual problem is?

**Comment by maxsamukha** [comment]
> I have a lot of experience with this. Please open a new report. The problem with reopening with something else is we lose the correspondence between issues and PRs. Also, what may appear to be the same problem may not actually be the same one.

My experience is non-existent compared to yours. I can hardly walk yet. Anyway, this time you are right. It must be a duplicate of multiple other reports: https://issues.dlang.org/show_bug.cgi?id=23646

**Comment by WalterBright** [comment]
@maxsamukha  Thanks for isolating it and filing the report. I'll look into it after this one is fixed.

**Comment by WalterBright** [comment]
Spent several hours boiling the problem down to this. Couldn't make it smaller.
```
template isCallable(alias callable)
{
    static if (is(typeof(&callable!())))
        enum bool isCallable = isCallable!(typeof(&callable!()));
    else
        enum bool isCallable = true;
}

string foo();

template FunctionTypeOf(alias func)
if (isCallable!func)
{
    alias FunctionTypeOf = typeof(foo);
}

template ReturnType(alias func)
{
    static if (is(FunctionTypeOf!func R == return))
        alias ReturnType = R;
}

template isAttrRange()
{
    alias NameType  = ReturnType!((string r) => r);
    pragma(msg, "isAttrRange ", NameType, " ", string);
    static assert(is(NameType == string));

    enum isAttrRange = is(NameType == string);
}

static assert(isAttrRange!());
```
Looks like I'll have to instrument the compiler next to see why `string != string`.

**Comment by WalterBright** [comment]
I spun off https://issues.dlang.org/show_bug.cgi?id=23651 which is blocking this PR. The problem is different from the one this PR is trying to solve.

**Comment by maxsamukha** [comment]
See also https://issues.dlang.org/show_bug.cgi?id=23652

**Review by RazvanN7 — COMMENTED** [review]
What we are hitting here is a general problem: how are static if declarations supposed to be handled? The answer seems to be: before everything else. If that is the case, maybe we should strive to have a more general solution (a visitor that first looks for static if declarations, or maybe store static if declarations at parse time in a different tree) than just introducing special cases to fix particular bugs.

The more we add ad-hoc fixes the more complicated it's going to be to fix the general case.

**Comment by WalterBright** [comment]
We've struggled with that problem for a long time. There's no obvious solution that I can think of. In the meantime, we can't let perfection get in the way of progress. At the very least, we are building a test suite of cases that we need to work, to test against any new ideas.

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**